tls/s2n_server_hello.c

```diff
@@ -189,11 +208,29 @@ int s2n_server_hello_recv(struct s2n_connection *conn)
     return 0;
 }
 
+int s2n_server_hello_write_message(struct s2n_connection *conn)
+{
+    /* The actual_protocol_version is set while processing the CLIENT_HELLO message, so
+     * it could be S2N_TLS13. SERVER_HELLO should always respond with the legacy version.
+     * https://tools.ietf.org/html/rfc8446#section-4.1.3 */
+    uint16_t legacy_protocol_version = MIN(conn->actual_protocol_version, S2N_TLS12);
+    uint8_t protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN];
+    protocol_version[0] = (uint8_t)(legacy_protocol_version / 10);
+    protocol_version[1] = (uint8_t)(legacy_protocol_version % 10);
+
+    GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
+    GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, conn->session_id_len));
+    GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, conn->session_id, conn->session_id_len));
+    GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, conn->secure.cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN));
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, S2N_TLS_COMPRESSION_METHOD_NULL));
+
+    return 0;
+}
+
 int s2n_server_hello_send(struct s2n_connection *conn)
 {
```

Comment:
as you are in here:

```
notnull_check(conn);
```

i think it is wise to include these anywhere we dereference a pointer argument.

Reply:
I updated the existing functions to include the check as well.